### PR TITLE
Fix --offline flag

### PR DESCRIPTION
**Pull request by cbclemmer**
---
Please include the following checklist in your PR:

* [x] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#conventions).
* [x] Any changes that could be relevant to users [have been recorded in the changelog](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#changelog).
* [x] The documentation has been updated, if necessary.

The --offline flag previously created in #2578 but was only implemented for the install command even thought the flag didn't throw an error whenever the build command was run. This PR adds functionality for the --offline flag with the build command.
Additionally there is a new PackageTest for the flag using the build command.

### CODE REVIEW

**Comment by ulysses4ever** [comment]
Presumably, fixes #5346

**Comment by cbclemmer** [comment]
@ulysses4ever Yeah, I was waiting until I got the tests passing before I pinged anyone

**Comment by ulysses4ever** [comment]
No worries! Let's see...

**Comment by cbclemmer** [comment]
@ulysses4ever I got all the test steps you were talking about before working. The only issue is that when it errors it gives the file the error was called from and that could change later on. Not sure what the best fix for this is because I just sent up the error like `rebuildTargets` is supposed to do by defining an error `BuildOutcome` object. 

Also as a note, because it's not downloading it from a remote repo I have to test the `PackageLocation` to be an instance of `RepoTarballPackage`, `RemoteTarballPackage`, or `RemoteSourceRepoPackage` instead of testing whether the `BuildStatus` is `BuildStatusDownload` like in the `asyncDownloadPackages` function (The test is of type `RepoTarballPackage`).

**Comment by cbclemmer** [comment]
@ulysses4ever I can probably create a tag like there is for the ghc version etc. to capture error details and that would fix it. I don't think there is anything like that at the moment.

**Comment by ulysses4ever** [comment]
I'm sorry I've been running a bit behind on cabal things. I'd have to defer this choice to someone more knowledgeable or just trust you.

**Comment by Mikolaj** [comment]
@cbclemmer: I'm thinking, could you have a look at `git blame` or PR authors of the related code and ping a few previous contributors (unless they are from > 5 years ago, I guess)? If nothing comes out of it, I will try to understand it, advise, merge and we'd depend on feedback from outraged users instead. ;)

**Comment by cbclemmer** [comment]
@ulysses4ever No worries. I just had an idea and wanted to share. I'll look around the output validator code and see what I can find.

**Comment by cbclemmer** [comment]
@ulysses4ever I added the tag and now the tests pass. It looks like it was several years ago since the file that normalizes the output was last edited.

**Review by Kleidukos — CHANGES_REQUESTED** [review]
I guess we can do without the .vscode/tasks.json ;)

File: cabal-install/src/Distribution/Client/ProjectBuilding.hs
Comment (ulysses4ever) [inline cabal-install/src/Distribution/Client/ProjectBuilding.hs:608]: Is it possible to use a guard instead of this `if`? I hope this would make the diff much better (the old code won't have to be bumped to the right).

File: cabal-testsuite/src/Test/Cabal/OutputNormalizer.hs
Comment (ulysses4ever) [inline cabal-testsuite/src/Test/Cabal/OutputNormalizer.hs:30]: This change needs a comment just like most of the lines above and below. And probably a more expanded comment somewhere else (e.g. commit message and/or PR description). It looks fishy tbh (I'm happy to be convinced otherwise, but again, please provide an explanation for it). The root of the problem, I think, is in the code where you call `error`: I don't believe it should be done that way.

File: cabal-testsuite/PackageTests/OfflineFlag/offlineFlag.out
Comment (ulysses4ever) [inline cabal-testsuite/PackageTests/OfflineFlag/offlineFlag.out:10]: Why do we get an exception? This shouldn't be the case, I think. If there's a remote dependency that is not in the cache and we received `--offline`, we need to gracefully shutdown printing a helpful message. The message is already up (see the line above). But the shutdown shouldn't bring up a stack trace!

File: cabal-install/src/Distribution/Client/ProjectBuilding.hs
Comment (ulysses4ever) [inline cabal-install/src/Distribution/Client/ProjectBuilding.hs:661]: How about:
```suggestion
        makeError n v = "--offline was specified, hence refusing to download the package "
```

File: cabal-install/src/Distribution/Client/ProjectBuilding.hs
Comment (ulysses4ever) [inline cabal-install/src/Distribution/Client/ProjectBuilding.hs:658]: I think it shouldn't involve `error`: it's not an exceptional situation like a network failure, or something else out of our control, it's we who are refusing to do something because of the flag -- it's no exception! So, I think, it should be `DependentFailed` rather than `DownloadFailed`. Perhaps someone has another opinion.

**Review by ulysses4ever — COMMENTED** [review]
We are getting there!

Also, what's up with `cabal-testsuite/PackageTests/MultipleLibraries/T7270/p/Main`?

**Comment by ulysses4ever** [comment]
@cbclemmer just a gentle ping. Great work here, just need to fix minor things...

File: cabal-install/src/Distribution/Client/ProjectBuilding.hs
Comment (cbclemmer) [reply cabal-install/src/Distribution/Client/ProjectBuilding.hs]: Moved to a guard at the beginning of the function

File: cabal-install/src/Distribution/Client/ProjectBuilding.hs
Comment (cbclemmer) [reply cabal-install/src/Distribution/Client/ProjectBuilding.hs]: `DependentFailed` needs another package as a parameter this looks like it's used to say that the package that is failing has a dependency that failed. I need to say that this specific package failed while also not producing an exception. Nothing else in the `BuildFailureReason` data type looked like it suited the problem so I added a `GracefulFailure` option to `BuildFailureReason` and it takes a string that will be printed to the terminal without any extra text or callstack, I think anything more specific would clutter up the data type more than is needed.

File: cabal-install/src/Distribution/Client/ProjectBuilding.hs
Comment (cbclemmer) [reply cabal-install/src/Distribution/Client/ProjectBuilding.hs]: Changed

File: cabal-testsuite/PackageTests/OfflineFlag/offlineFlag.out
Comment (cbclemmer) [reply cabal-testsuite/PackageTests/OfflineFlag/offlineFlag.out]: As stated above, I changed the build failure type to take a custom string and not produce an exception.

File: cabal-testsuite/src/Test/Cabal/OutputNormalizer.hs
Comment (cbclemmer) [reply cabal-testsuite/src/Test/Cabal/OutputNormalizer.hs]: This is now removed because the build does not produce an exception.

**Comment by cbclemmer** [comment]
@ulysses4ever My bad. I forgot to get back to this PR. I've replied to your code review items with updates. I'll ping you again when the CI passes.

**Comment by Kleidukos** [comment]
We really ought to provide scripts that run the validators locally

**Comment by cbclemmer** [comment]
@Kleidukos There is a `validate.sh` script , but I can't ever get the configuration to be right on my machine. A docker image would probably be the easiest thing to set up.

**Comment by ulysses4ever** [comment]
I sense we're getting close. We need reviewers and possibly a backport. I'll add my final review ASAP but we'd need another one.

**Comment by cbclemmer** [comment]
@Kleidukos I didn't see your review before. It's deleted now. I must have randomly pressed F5 at some point...

**Comment by Kleidukos** [comment]
> @Kleidukos I didn't see your review before. It's deleted now. I must have randomly pressed F5 at some point...

No problem :)

**Comment by ulysses4ever** [comment]
@Kleidukos you requested changes on this PR (https://github.com/haskell/cabal/pull/8676#pullrequestreview-1275624602) and I believe they were applied, so maybe you could either unblock it or even approve it?

**Review by Kleidukos — APPROVED** [review]
LGTM! :)

File: cabal-install/src/Distribution/Client/ProjectBuilding.hs
Comment (ulysses4ever) [inline cabal-install/src/Distribution/Client/ProjectBuilding.hs:589]: I think you could use `notNull` instead of `not` and `null` separately.

File: cabal-install/src/Distribution/Client/ProjectBuilding.hs
Comment (ulysses4ever) [reply cabal-install/src/Distribution/Client/ProjectBuilding.hs]: Full disclosure: I'm no expert in Cabal error types. But, I thought that `DependentFailed` would work quite all right for this use case. Imagine you depend on 10 packages and 9 of them are in the cache. Then, you may want to know which one wasn't there. On the other hand, if there're two dependencies that are not in the cache then reporting just one is not so nice, perhaps. Although, I could live with reporting one that happened to be the first to be required during the build…

That said, I don't have a strong preference. I'm just not a big fan of changing data types that sound fundamental. Cabal is a library and this would be a breaking change...

**Review by ulysses4ever — APPROVED** [review]

**Comment by ulysses4ever** [comment]
@cbclemmer you should feel free to put the squash+merge label now. 

Thank you for your contribution, and keep the good stuff coming!!! :tada:

**Comment by cbclemmer** [comment]
@ulysses4ever Thanks, I looked up notNull on hoogle and I thought it was included in the prelude but apparently it wasn't. I've reverted my changes and now it needs to do the build process again. Thanks for reviewing it

**Comment by ulysses4ever** [comment]
@Mergify backport 3.10

**Pull request merged**
